KAFKA-15950: Serialize broker heartbeat requests

soarez opened the pull request:
In between HeartbeatRequest being sent and the response being handled, i.e. while a HeartbeatRequest is in flight, an extra request may be immediately scheduled if propagateDirectoryFailure, setReadyToUnfence, or beginControlledShutdown is called.

To prevent the extra request, we can avoid the extra requests by checking whether a request is in flight, and delay the scheduling if necessary.

Some of the tests in BrokerLifecycleManagerTest are also improved to remove race conditions and reduce flakiness.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

soarez commented:
@junrao: please take a look

junrao commented on core/src/main/scala/kafka/server/BrokerLifecycleManager.scala:
```diff
@@ -166,6 +166,19 @@ class BrokerLifecycleManager(
   private var registered = false
 
+  /**
+   * True if a request has been sent and a response or timeout has not yet been processed.
```
a request => a Heartbeat request ?

soarez replied:
It applies to both Heartbeat and Registration requests which are the only requests this manager sends.

I think we're also fixing a bug were a second registration request could have been scheduled after a call to `propagateDirectoryFailure` and before the registration response is received.

junrao replied:
Hmm, if that's the case, should we set communicationInFlight to true when sending the very first Registration request?

soarez replied:
We do. The very first Registration request is also scheduled via a `CommunicationEvent`. 
`communicationInFlight = true` is now set after both `_channelManager.sendRequest(new BrokerRegistrationRequest...` and `_channelManager.sendRequest(new BrokerHeartbeatRequest...`

junrao commented on core/src/main/scala/kafka/server/BrokerLifecycleManager.scala:
```diff
@@ -366,8 +379,27 @@ class BrokerLifecycleManager(
+
+  private class BrokerRegistrationResponseEvent(response: ClientResponse, timedOut: Boolean) extends EventQueue.Event {
+    override def run(): Unit = {
+      communicationInFlight = false
```
It seems that we don't need this since `communicationInFlight` is only set to true when sending a `HeartbeatRrequest`?

soarez replied:
I don't think so. We set `communicationInFlight = true` in `CommunicationEvent.run`, which sends both heartbeats but also registration requests.

Maybe I'm missing something?

junrao replied:
Since `NodeToControllerChannelManagerImpl` uses 1 for `maxInFlightRequestsPerConnection`, we could also add a `sendRequest` api in `NodeToControllerChannelManager`, instead 

networkClient.ready(node, now)

junrao replied:
Ok. I wasn't sure that communicationInFlight is for both requests.

Since `NodeToControllerChannelManagerImpl` uses 1 for `maxInFlightRequestsPerConnection`, another option is to add a `canSendRequest` api in `NodeToControllerChannelManager`, instead maintaining `communicationInFlight` here. 

`canSendRequest` api can be implemented through `networkClient.ready(node, now)`.

junrao commented on core/src/main/scala/kafka/server/BrokerLifecycleManager.scala:
```diff
@@ -453,79 +490,73 @@ class BrokerLifecycleManager(
-          scheduleNextCommunicationAfterSuccess()
-      }
+  private def scheduleNextCommunicationImmediately(): Unit = {
+    if (communicationInFlight) {
```
I am not sure if this completely avoids duplicated `HeartbeatRequest`s. Consider the following flow. The event queue contains a `BrokerHeartbeatResponseEvent` followed by an `OfflineDirEvent`. The event queue thread processes the former.  It sets `communicationInFlight` to false and enqueues a `CommunicationEvent`. It then process the `OfflineDirEvent`. Since `communicationInFlight` is false, it enqueues another `CommunicationEvent`. Now, there are two `CommunicationEvent`s and each will introduce a separate `HeartbeatRequest`.

soarez replied:
You are correct, thanks for finding this. I placed the `communicationInFlight` check in `scheduleNextCommunicationImmediately`, but `BrokerHeartbeatResponseEvent` bypasses that by calling `scheduleNextCommunication` directly.

I'm moving the check to `scheduleNextCommunication` so we can catch all cases. Since `CommunicationEvent` is only scheduled from this function, we can make sure to not schedule it when a request is in flight.

junrao commented on core/src/test/scala/unit/kafka/server/BrokerLifecycleManagerTest.scala:
```diff
@@ -197,11 +197,14 @@ class BrokerLifecycleManagerTest {
-      context.poll()
+  def poll[T](ctx: RegistrationTestContext, manager: BrokerLifecycleManager, future: Future[T]): T = {
+    while (ctx.mockChannelManager.unsentQueue.isEmpty) {
+      if (manager.eventQueue.isEmpty)
```
Why do we need this check? If the eventQueue has a event scheduled event at a future time, we need to advance the time to drain the event, right?

soarez replied:
You're right, this is incorrect. We must only advance the time if the eventQueue has an event scheduled at a future time, unless the next event is a timeout event.

junrao reviewed (commented):
@soarez : Thanks for the PR. Left a few comments.

cmccabe commented:
Thanks for looking at this, @soarez .

I don't think `nextSchedulingShouldBeImmediate` should be reset to false until the `CommunicationEvent` is run.

I also think we should be a bit more careful about setting `communicationInFlight` ... it should only be set after `_channelManager.sendRequest` has been called. And even that call should probably be wrapped in a `try...catch`. If `sendRequest` throws, we don't want to be stuck never sending another request.

(I don't think sendRequest is supposed to throw, but we should be careful.)

I don't think prepend is needed anywhere. To be honest, the presence of `prepend` usually indicates a bug... :) Instead, `CommunicationEvent.run()` should gracefully handle running when `communicationInFlight = true` Probably by doing nothing.

cmccabe commented:
> Instead, CommunicationEvent.run() should gracefully handle running when communicationInFlight = true Probably by doing nothing.

This also implies that `scheduleNextCommunicationAfterSuccess` should be checking `nextSchedulingShouldBeImmediate` (not sure why it doesn't already)

Probably `nextSchedulingShouldBeImmediate` should be renamed to something like "dirty" since it doesn't actually mean the next scheduling should always be immediate. Like if there has been a communication error (`scheduleNextCommunicationAfterFailure`) we do not want to immediately reschedule, no matter what.

soarez commented:
@cmccabe thanks for having a look. 

I've made the following changes:

* Replaced the use of `prepend` with `append` in `propagateDirectoryFailure`
* Moved `communicationInFlight = true` after `_channelManager.sendRequest()`
* Always set `nextSchedulingShouldBeImmediate = false` in `scheduleNextCommunicationAfterFailure`
* Moved checking if `communicationInFlight = true` to `CommunicationEvent.run()`

@junrao @cmccabe: please take another look

junrao commented on core/src/main/scala/kafka/server/BrokerLifecycleManager.scala:
```diff
@@ -166,6 +166,19 @@ class BrokerLifecycleManager(
+   * True if a request has been sent and a response or timeout has not yet been processed.
+   * This variable can only be read or written from the event queue thread.
+   */
+  private var communicationInFlight = false
```
The `NetworkClient` in `_channelManager` already maintains the state for in flight requests through `networkClient.ready(node, now)`. So, I am wondering if we need to maintain the state here. Could we just expose the state in `NetworkClient` through `_channelManager`?

soarez replied:
I don't see a straightforward way to do this.

Maybe I'm missing something, but`NetworkClient` does not look like a good place for this. `kafka.server.NodeToControllerChannelManagerImpl#sendRequest` isn't synchronous, it runs `requestThread.enqueue`, so if we rely on `networkClient` state we open the door to race conditions.
Also, `NodeToControllerRequestThread` can reset the network client at any point.

junrao replied:
Yes, I agree that it's not straightforward. We can just leave the code as it is.

junrao commented on core/src/test/scala/unit/kafka/server/BrokerLifecycleManagerTest.scala:
```diff
@@ -197,11 +197,14 @@ class BrokerLifecycleManagerTest {
       manager.eventQueue.wakeup()
-      context.time.sleep(5)
+    }
+    while (!future.isDone || ctx.mockClient.hasInFlightRequests) {
```
Do we need `ctx.mockClient.hasInFlightRequests`? It seems that `!future.isDone` is enough?

soarez replied:
Yes, it seems that is enough. Changing it.

junrao commented on server-common/src/main/java/org/apache/kafka/queue/KafkaEventQueue.java:
```diff
@@ -513,4 +513,28 @@ public void close() throws InterruptedException {
+     * Useful for unit tests, where we need to speed the clock up until
+     * deferred events are ready to run.
+     */
+    public Object pendingDeferredEvent() {
```
Could we return Optional instead of null?

soarez replied:
Sure 👍

junrao commented on server-common/src/main/java/org/apache/kafka/queue/KafkaEventQueue.java:
```diff
@@ -513,4 +513,28 @@ public void close() throws InterruptedException {
+    public Object pendingDeferredEvent() {
+        lock.lock();
+        try {
+            if (eventHandler.head.next != eventHandler.head) {
```
Deferred events are not stored under `head`. It seems there is not need to check `head`?

soarez replied:
Well, yes, deferred events are not stored under `head`, but if there are any events under `head` we should not fast forward the clock, as those will run regardless. The point here is to return the first event that will run next, when its deadline comes, but if and only if there's nothing else to do until then. I'll get a better name and javadoc for this method so this makes more sense.

junrao commented on server-common/src/main/java/org/apache/kafka/queue/KafkaEventQueue.java:
```diff
@@ -513,4 +513,28 @@ public void close() throws InterruptedException {
+                return null;
+            }
+            EventContext eventContext = entry.getValue();
+            if (eventContext.insertionType == EventInsertionType.DEFERRED) {
```
Do we need this check? It seems only `DEFERRED` events are stored in `deadlineMap`.

soarez replied:
That is incorrect. Events of any type can end up in `deadlineMap` as long as they have a deadline specified.

The notion of deadline here is a bit confusing. A deadline for events of type `DEFERRED` is the time after which the event should run. For events of other types, the deadline is the time after which `TimeoutException` should be delivered if the event hasn't run yet. So we want to ignore events that are not of type `DEFERRED`.

junrao reviewed (commented):
@soarez : Thanks for the updated PR. A few more comments.

junrao commented on core/src/main/scala/kafka/server/BrokerLifecycleManager.scala:
```diff
@@ -364,10 +377,30 @@ class BrokerLifecycleManager(
+  // to continue handling the response on the event handler thread
   private class BrokerRegistrationResponseHandler extends ControllerRequestCompletionHandler {
     override def onComplete(response: ClientResponse): Unit = {
+      eventQueue.prepend(new BrokerRegistrationResponseEvent(response, false))
```
@cmccabe : Do you have concerns on the usage of `prepend` here?

junrao reviewed (commented):
@soarez : Thanks for the updated PR. Left a comment for Colin to check. Are the test failures related to this PR?

rondagostino commented:
`51 tests have failed.  There are 49 new tests failing, 2 existing failing`

This seems like a lot more failures than we are used to recently.  @OmniaGM Can you rebase this onto latest `trunk` to see if we can get a cleaner build?

OmniaGM commented:
> `51 tests have failed. There are 49 new tests failing, 2 existing failing`
> 
> This seems like a lot more failures than we are used to recently. @OmniaGM Can you rebase this onto latest `trunk` to see if we can get a cleaner build?

done! I don't have related tests failing locally. Waiting for the build now.

rondagostino commented:
`12 tests have failed.  There are 12 new tests failing, 0 existing failing`

I think this PR can be merged now assuming @cmccabe takes a look at the question above regarding prepend to event queue.

soarez committed:
KAFKA-15950: Serialize heartbeat requests

In between HeartbeatRequest being sent and the response being handled,
i.e. while a HeartbeatRequest is in flight, an extra request may be
immediately scheduled if propagateDirectoryFailure, setReadyToUnfence,
or beginControlledShutdown is called.

To prevent the extra request, we can avoid the extra requests by checking
whether a request is in flight, and delay the scheduling if necessary.

Some of the tests in BrokerLifecycleManagerTest are also improved to
remove race conditions and reduce flakiness.

soarez commented:
Rebased this due to merge conflict. 

@gaurav-narula I also made what I believe are some improvement to `BrokerLifecycleManagerTest.testKraftJBODMetadataVersionUpdateEvent`, can you please take a look?

junrao commented on core/src/main/scala/kafka/server/BrokerLifecycleManager.scala:
```diff
@@ -551,9 +580,11 @@ class BrokerLifecycleManager(
 
   private def scheduleNextCommunication(intervalNs: Long): Unit = {
-    trace(s"Scheduling next communication at ${MILLISECONDS.convert(intervalNs, NANOSECONDS)} " +
+    val nanos = if (nextSchedulingShouldBeImmediate) 0 else intervalNs
```
nanos =>  adjustedIntervalNs?

junrao commented on server-common/src/main/java/org/apache/kafka/queue/KafkaEventQueue.java:
```diff
@@ -513,4 +514,36 @@ public void close() throws InterruptedException {
+     * This is useful for unit tests, where to make progress, we need to
+     * speed the clock up until the next scheduled event is ready to run.
+     */
+    public Optional<Event> scheduledAfterIdling() {
```
scheduledAfterIdling => firstDeferredIfIdling ?

junrao commented on core/src/test/scala/unit/kafka/server/BrokerLifecycleManagerTest.scala:
```diff
@@ -197,11 +197,17 @@ class BrokerLifecycleManagerTest {
+    while (ctx.mockChannelManager.unsentQueue.isEmpty) {
+      // If the manager is idling until scheduled events we need to advance the clock
+      if (manager.eventQueue.scheduledAfterIdling()
+        .filter(!_.getClass.getSimpleName.endsWith("TimeoutEvent")) // avoid triggering timeout events
```
I couldn't find a `TimeoutEvent` class.

soarez replied:
This is a bit ugly. It only applies to `RegistrationTimeoutEvent`. We want to avoid triggering those deferred events.

Instead of checking for that class specifically, I thought there may be other timeout event types in the future and we'll also want to avoid those. I also ruled out extending EventQueue.Event as we only care about this in this test suite.

I feel ambivalent about the approach I've taken here, so please let me know if you have a preference.

junrao replied:
I see. Perhaps we could add a step after `poll(ctx, manager, registration)` to explicitly drain `RegistrationTimeoutEvent` first? This makes the test safer since it won't be sensitive to the timeout config values.

soarez replied:
That's a good idea. I'm making that change.

junrao commented on core/src/test/scala/unit/kafka/server/BrokerLifecycleManagerTest.scala:
```diff
@@ -219,15 +225,16 @@ class BrokerLifecycleManagerTest {
     manager.propagateDirectoryFailure(Uuid.fromString("h3sC4Yk-Q9-fd0ntJTocCA"))
+    assertEquals(Set("h3sC4Yk-Q9-fd0ntJTocCA"), nextHeartbeatDirs())
     manager.propagateDirectoryFailure(Uuid.fromString("ej8Q9_d2Ri6FXNiTxKFiow"))
+    assertEquals(Set("h3sC4Yk-Q9-fd0ntJTocCA", "ej8Q9_d2Ri6FXNiTxKFiow"), nextHeartbeatDirs())
```
It seems this could still be flaky. 

`poll(ctx, manager, registration)
`
A CommunicationEvent c1 with a delay of 0 is scheduled in manager.eventQueue.

`assertEquals(Set.empty, nextHeartbeatDirs())
`
c1 is processed in `ctx.mockChannelManager` and a BrokerHeartbeatResponseEvent b1 is added to `manager.eventQueue`. b1 is processed at `manager.eventQueue` and a CommunicationEvent c2 with a delay of 100 is scheduled in `manager.eventQueue`.

`manager.propagateDirectoryFailure(Uuid.fromString("h3sC4Yk-Q9-fd0ntJTocCA"))
`
An OfflineDirEvent o1 is appended to `manager.eventQueue`, but not yet processed.

`assertEquals(Set("h3sC4Yk-Q9-fd0ntJTocCA"), nextHeartbeatDirs())
`
nextHeartbeatDirs() causes the time to advance quickly and passes 100. c2 is processed at `manager.eventQueue` and a HeartBeat request h1 is added to `ctx.mockChannelManager`.

o1 is processed at `manager.eventQueue` and a CommunicationEvent c3 is appended to `manager.eventQueue`.  c3 is processed at `manager.eventQueue` sets `nextSchedulingShouldBeImmediate` to true.

h1 is processed by `ctx.mockChannelManager` and adds BrokerHeartbeatResponseEvent b2 to `manager.eventQueue`. `manager.eventQueue` processes b2 and schedules a CommunicationEvent c4 with a delay of 0 in `manager.eventQueue`. c4 is processed at `manager.eventQueue` and a HeartBeat request h2 (doesn't pick up ej8Q9_d2Ri6FXNiTxKFiow) is added to `ctx.mockChannelManager`.

`manager.propagateDirectoryFailure(Uuid.fromString("ej8Q9_d2Ri6FXNiTxKFiow"))`
`assertEquals(Set("h3sC4Yk-Q9-fd0ntJTocCA", "ej8Q9_d2Ri6FXNiTxKFiow"), nextHeartbeatDirs())`
Now the above assertion will fail since `nextHeartbeatDirs()` will pick up c4 which doesn't include ej8Q9_d2Ri6FXNiTxKFiow.

soarez replied:
Thanks for carefully thinking this through. 

I believe there's a mistake in that sequence of events though:

> nextHeartbeatDirs() causes the time to advance quickly and passes 100

This isn't how `BrokerLifecycleManagerTest#poll()` works. It will only advance time if there is no event in the queue that can immediately run, so o1 will always run before c2.

> o1 is processed at manager.eventQueue and a CommunicationEvent c3 is appended to manager.eventQueue

Because o1 runs before c2, the EventQueue will override c2 with c3 as they're both deferred events with the same tag.

After running this test tens of thousands of iterations, I was unable to reproduce a flaky run.

junrao replied:
Thanks for the explanation. You are right. So, this is not an issue.

junrao commented on core/src/test/scala/unit/kafka/server/BrokerLifecycleManagerTest.scala:
```diff
@@ -254,33 +261,38 @@ class BrokerLifecycleManagerTest {
-    }
 
+    def doPoll[T<:AbstractRequest](response: AbstractResponse) = poll(ctx, manager, prepareResponse[T](ctx, response))
+    def nextRequest() = doPoll[AbstractRequest](new BrokerHeartbeatResponse(new BrokerHeartbeatResponseData()))
```
nextRequest => nextHeartbeatRequest ?

junrao commented on core/src/test/scala/unit/kafka/server/BrokerLifecycleManagerTest.scala:
```diff
@@ -254,33 +261,38 @@ class BrokerLifecycleManagerTest {
-    }
+
+    // We may have to accept some heartbeats before the new registration is sent
+    while (nextRequest().isInstanceOf[BrokerHeartbeatRequest])()
```
Hmm, the next request could be `BrokerRegistration`. In that case, it would be incorrect to mock a `BrokerRegistrationResponse`, right?

soarez replied:
I found it difficult to predict when the new `BrokerRegistration` request is sent. Depending on scheduling it could be sent straight away or it could get queued behind another heartbeat.

It is indeed incorrect to mock `BrokerRegistrationResponse` for a `BrokerRegistration` request. It is however harmless because `BrokerRegistrationResponseEvent` will `scheduleNextCommunicationAfterFailure()` if `!response.responseBody().isInstanceOf[BrokerRegistrationResponse]`.

And so I think this is the simplest and most straightforward way to deal with the potential extra heartbeat request - by consuming the requests until they're no longer `BrokerHeartbeatRequest`, even if that implies failing the first attempt at sending `BrokerRegistrationRequest`.

I've added comment to explain this. Let me know what you think.

junrao replied:
`prepareResponse` knows the request type. Could we generate the response corresponding to the request type there?

soarez replied:
No, I don't think so. `prepareResponse`  delegates to `MockClient` which expects a predetermined request-response order. It supports a  `RequestMatcher` which `prepareResponse` uses to extract the request, but it does not support preparing a conditional response. We need a larger change to add support in `MockClient` for conditional prepared responses.

junrao reviewed (commented):
@soarez : Thanks for the updated PR. A few more comments.

soarez commented:
Thanks for the review @junrao. Please take another look.

junrao reviewed (commented):
@soarez : Thanks for the updated PR. Just a couple of more comments.

junrao commented on core/src/test/scala/unit/kafka/server/BrokerLifecycleManagerTest.scala:
```diff
@@ -254,33 +262,41 @@ class BrokerLifecycleManagerTest {
-      assertEquals(1200, manager.brokerEpoch)
-    }
+
+    // Depending on scheduling, the next request could either be BrokerRegistration or BrokerHeartbeat.
```
Before calling `manager.handleKraftJBODMetadataVersionUpdate()`, there should be 1 `CommunicationEvent` with a delay of 100 in `eventQueue`. Until we call `poll`, this `CommunicationEvent` will remain in `eventQueue`. Calling `manager.handleKraftJBODMetadataVersionUpdate()` causes a `KraftJBODMetadataVersionUpdateEvent` with no delay to be added to `eventQueue`. The `KraftJBODMetadataVersionUpdateEvent` will then be processed and replace the `CommunicationEvent` with a delay of 0. The `CommunicationEvent` will then be processed, which causes a `BrokerRegistration` request to be sent. So, it seems that the next request should always be `BrokerRegistration` when we call `poll`?

soarez replied:
You're right. This does not seem necessary. I removed it.

junrao reviewed (commented):
@soarez : Thanks for the updated PR. Just one more question.

Also, are the test failures related? If not, it would be useful to file jiras if they are not tracked already.

junrao reviewed (approved):
@soarez : Thanks for the latest update. The code LGTM. Are the test failures related? Also, jdk 17 didn't complete. Perhaps you could trigger the test again?

soarez commented:
Hey @junrao . Sorry I didn't reply about the failing tests earlier, I was still looking into it and should've made that clear.

I had a look at the test results. It's strange that all the failed tests are categorized as "new failures", but most of them are known flaky tests.

These tests already have a respective open JIRA:

* KAFKA-8041 kafka.server.LogDirFailureTest.testIOExceptionDuringCheckpoint(String).quorum=kraft
* KAFKA-8115 org.apache.kafka.trogdor.coordinator.CoordinatorTest.testTaskRequestWithOldStartMsGetsUpdated()
* KAFKA-8250 kafka.api.DelegationTokenEndToEndAuthorizationWithOwnerTest.testNoConsumeWithDescribeAclViaAssign(String).quorum=kraft
* KAFKA-8677  kafka.api.DelegationTokenEndToEndAuthorizationWithOwnerTest."testNoDescribeProduceOrConsumeWithoutTopicDescribeAcl(String, boolean).quorum=kraft, isIdempotenceEnabled=true"
* KAFKA-13514 org.apache.kafka.clients.consumer.StickyAssignorTest.testLargeAssignmentAndGroupWithNonEqualSubscription(boolean).hasConsumerRack = false
* KAFKA-15104 org.apache.kafka.tools.MetadataQuorumCommandTest.testDescribeQuorumReplicationSuccessful [1] Type=Raft-Combined, MetadataVersion=3.8-IV0, Security=PLAINTEXT
* KAFKA-15411 kafka.api.DelegationTokenEndToEndAuthorizationWithOwnerTest."testNoProduceWithDescribeAcl(String, boolean).quorum=kraft, isIdempotenceEnabled=true"
* KAFKA-15898 org.apache.kafka.controller.QuorumControllerTest.testFenceMultipleBrokers()
* KAFKA-15914 org.apache.kafka.connect.integration.OffsetsApiIntegrationTest.testAlterSinkConnectorOffsetsOverriddenConsumerGroupId
* KAFKA-15917 OffsetsApiIntegrationTest.testAlterSinkConnectorOffsetsZombieSinkTasks
* KAFKA-15927 org.apache.kafka.connect.mirror.integration.IdentityReplicationIntegrationTest.testReplicateSourceDefault()
* KAFKA-15928 org.apache.kafka.controller.QuorumControllerTest.testBalancePartitionLeaders()
* KAFKA-15940 kafka.server.DynamicBrokerReconfigurationTest.testTrustStoreAlter(String).quorum=kraft
* KAFKA-15945 org.apache.kafka.connect.mirror.integration.IdentityReplicationIntegrationTest.testSyncTopicConfigs()
* KAFKA-16024 kafka.api.SaslPlainPlaintextConsumerTest.testCoordinatorFailover(String, String).quorum=kraft+kip848.groupProtocol=consumer
* KAFKA-16174 org.apache.kafka.tools.MetadataQuorumCommandTest.testDescribeQuorumStatusSuccessful [5] Type=Raft-Combined, MetadataVersion=3.8-IV0, Security=PLAINTEXT
* KAFKA-16225 kafka.server.LogDirFailureTest.testIOExceptionDuringLogRoll(String).quorum=kraft
* KAFKA-16323 kafka.server.ReplicaManagerTest.testRemoteFetchExpiresPerSecMetric()

I could not find JIRAs for these, so I created them:

* KAFKA-16402 org.apache.kafka.controller.QuorumControllerTest.testSnapshotSaveAndLoad()
* KAFKA-16403 org.apache.kafka.streams.examples.wordcount.WordCountDemoTest.testCountListOfWords()
* KAFKA-16404 org.apache.kafka.streams.examples.wordcount.WordCountDemoTest.testGetStreamsConfig()

I also couldn't find JIRAs for these, but they're timeouts, so it could have been a failure of the test environment and not necessarily something wrong in the test. I've rebased to trigger a new build, and we'll see if they remain.

* org.apache.kafka.clients.consumer.CooperativeStickyAssignorTest.testAssignmentAndGroupWithNonEqualSubscriptionNotTimeout(boolean).hasConsumerRack = false
* org.apache.kafka.clients.consumer.StickyAssignorTest.testAssignmentAndGroupWithNonEqualSubscriptionNotTimeout(boolean).hasConsumerRack = false
* org.apache.kafka.tools.consumer.group.DeleteOffsetsConsumerGroupCommandIntegrationTest.testDeleteOffsetsNonExistingGroup().
* org.apache.kafka.connect.integration.BlockingConnectorTest.testBlockInConnectorValidate
* org.apache.kafka.server.ClientMetricsManagerTest.testCacheEvictionWithMultipleClients()


None of the 26 tests is related to this change.

junrao commented:
@soarez : Are the failed tests in the latest run being tracked?

soarez commented:
@junrao all the failed tests are being tracked.

These were already tracked:

* KAFKA-8041 kafka.server.LogDirFailureTest.testIOExceptionDuringCheckpoint(String).quorum=kraft
* KAFKA-8115 org.apache.kafka.trogdor.coordinator.CoordinatorTest.testTaskRequestWithOldStartMsGetsUpdated()
* KAFKA-15772 org.apache.kafka.tiered.storage.integration.TransactionsWithTieredStoreTest.testAbortTransactionTimeout(String).quorum=kraft
* KAFKA-15897 kafka.server.ControllerRegistrationManagerTest.testWrongIncarnationId()
* KAFKA-15898 org.apache.kafka.controller.QuorumControllerTest.testFenceMultipleBrokers()
* KAFKA-15921 kafka.api.SaslScramSslEndToEndAuthorizationTest.testAuthentications(String).quorum=kraft
* KAFKA-15927 org.apache.kafka.connect.mirror.integration.IdentityReplicationIntegrationTest.testReplicateSourceDefault()
* KAFKA-15927 org.apache.kafka.connect.mirror.integration.MirrorConnectorsWithCustomForwardingAdminIntegrationTest.testReplicateSourceDefault()
* KAFKA-15928 org.apache.kafka.controller.QuorumControllerTest.testBalancePartitionLeaders()
* KAFKA-15945 org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationTransactionsTest.testSyncTopicConfigs()
* KAFKA-15961 kafka.controller.ControllerIntegrationTest.testTopicIdPersistsThroughControllerRestart()
* KAFKA-16225 kafka.server.LogDirFailureTest.testIOExceptionDuringLogRoll(String).quorum=kraft
* KAFKA-16323 kafka.server.ReplicaManagerTest.testRemoteFetchExpiresPerSecMetric()

This one was not, so I created a JIRA:

* KAFKA-16422 org.apache.kafka.controller.QuorumControllerMetricsIntegrationTest."testFailingOverIncrementsNewActiveControllerCount(boolean).true"

junrao commented:
@soarez : Thanks for triaging the test failures. Will merge the PR.
